### ISSUE TITLE
refactor/redesign stages of `flox activate`

### ISSUE THREAD

**Issue by limeytexan** [flox/flox#1187]
We have multiple issues related to the current version of `flox activate` that might warrant a re-think of our approach:

* #1186
* #1185
* #1174
* #1173
* #1135
* #1130 
* #1019

... and a few other that have been closed with "point" patches:

* #817 

I started looking into this in the context of https://github.com/flox/flox/pull/1176#issuecomment-2000261770 (portions of which to be copied below) and found that at a minimum we have an issue with the use of the term "non-interactive" in the code, which is confusing because it matches neither a) our previous use of the term nor b) the canonical definition of the word "interactive".

Logging this issue to track an holistic effort to address all known issues with a larger refactor/redesign.

List of tasks to be addressed, the details of which are described in the remainder of this issue:

- [ ] Proposal 1: rename/refactor our use of the terms "interactive", "non-interactive", and "in-place"
    - N.B. this naming has been introduced with a modification to `flox-activate.md` in #1176 
- [ ] Proposal 2: introduce the term "userShell" and precisely define how it is set as flox activate is invoked
- [ ] Proposal 3: review and potentially redesign flox activate along distinct stages of activation
- [ ] Proposal 4: improve automatic userShell detection (and/or specially handle the "direnv" case)

**Comment by limeytexan** [comment]
(By way of background, much of this append is derived from from https://github.com/flox/flox/pull/1176#issuecomment-2000261770 as noted above.)

### Proposal 1: rename/refactor our use of the terms "interactive", "non-interactive", and "in-place"

The rust codebase currently associates these terms with the following actions:

1. interactive: `flox activate`
1. non-interactive: `flox activate -- CMD`
1. in-place: `flox activate --print-script` or `flox activate` invoked with STDOUT redirected to something other than a TTY (e.g. `flox activate | cat`)

The canonical definition of "interactive" is the property of having STDOUT connected to a TTY. On this basis I'd argue we shouldn't be calling the second of these "non-interactive" because `flox activate -- bash` would definitely yield an interactive shell. We also previously defined these terms differently in [flox-bash](https://github.com/flox/flox-bash/blob/main/lib/commands/activate.sh#L31-L66), which is confusing.

To avoid further confusion I'd propose we refactor the code and documentation to refer to these [3] modes as:

1. **interactive**
1. **command** (formerly non-interactive with flox rust)
1. **in-place** (formerly non-interactive with flox bash)

### Proposal 2: introduce the term "userShell" and precisely define how it is set as `flox activate` is invoked

Define the term "userShell" as the user-specified shell that is used to determine:
1. the shell to be invoked in "interactive" mode
1. the language dialect to be used for "in-place" activations

The value of the "userShell" is set by one of three mechanisms, in this order of priority:
1. the `$FLOX_SHELL` environment variable if set will take precedence in defining the userShell
1. otherwise `flox` will detect the shell that is "listening" to its output by way of magic discussed/implemented in #1176 
1. and if that doesn't work then `flox` will fall back to using `$SHELL` from the environment

### Proposal 3: review and potentially redesign `flox activate` along distinct stages of activation

Thinking about my response to #1176 it dawned on me that there are parallels to be drawn between `flox activate` and the distinct stages of the Linux boot process. There are two properties of the Linux boot stages that I think are particularly relevant here:

1. each stage is responsible for preparing state used by the stage to follow
1. as each stage cedes control to the next, it does so by _replacing_ itself with the next stage (i.e. `kexec`)

It would be helpful to organise our thinking with these same properties in mind, and applying these concepts to the various modes of `flox activate` we have these mechanisms to work with:

1. stage 1: the (rust) `flox` CLI itself
    1. sets nix/flox-related constants in environment (e.g. `FLOX_VERSION`, `NIX_SSL_CERT_FILE`, etc.)
    1. sets other env variables `PATH`, `FLOX_ENV`, etc. 
    1. creates bash script to be consumed in "stage 2"
    1. creates scripts for _all known shell dialects_ to be used in "stage 3"
    1. finishes by `exec`ing the "stage 2" bash script (using a _flox-provided_ modern Nix `bash`) which inherits the variables from above
1. stage 2: a script invoked by `bash`
    1. invokes user-provided hook (written in bash)
    1. finishes by `exec`ing the "stage 3" command to be invoked, see below
1. stage 3 is where things get different depending on the activation mode and the userShell:
    1. "interactive": invoke `exec <userShell> <args>` with args that:
        1. defeat any sourcing of user-specific "dotfiles" that might mess things up (e.g. bash has a `--norc` option that is useful here)
        1. sources the appropriate language script created in step **1.d** above
    3. "in-place": outputs the appropriate script in the dialect of the userShell created in step **1.d** above 
    4. "command" mode: simply invoke `<cmd> <args>` as supplied by the user
        * **N.B.** no need to involve **userShell** in this mode at all

**Comment by mkenigs** [comment]
It seems to me that each of the 4 proposals are pretty distinct, so I would be in favor of closing this as it will be hard to discuss or prioritize as a group.

**Comment by limeytexan** [comment]
Thanks Matthew - I definitely want to understand your concern. The way I see it each of the four proposals is a different task in the summary above, and the reason that I presented them here in one place is both to record the related discussions in a single thread of conversation and because each one relies on the previous one(s) in turn. I certainly didn't expect to implement all four with one PR - it would definitely make more sense to implement each of these with a separate PRs linked to this one issue. Does that help clarify things?

**Comment by mkenigs** [comment]
I'm probably just not seeing the inter-dependencies you're seeing?

**Comment by limeytexan** [comment]
> I'm probably just not seeing the inter-dependencies you're seeing?

For example, I only came up with proposal 1 and 2 when I found it difficult to write up a description of proposal 3. Proposals 1 and 2 are really trivial in that regard, just clarifying terminology more than anything else.

Proposal 4 is admittedly a different thing, but one that came up in the context of the sorts of complexities that can arise in stage 3 so I thought it provided good context to help affirm the correctness of the 3-stage design. It's also less of a priority to be fair; if/when we get to the point of finishing 1-3 and want to copy 4 to a new issue and close this out then I'd be OK with that.  👍

**Comment by limeytexan** [comment]
Thanks @ysndr for your comments above.

> Proposal 3 is missing the distinctions between hook.on-activate and profile.common.
> hook.on-activate is called after the profile scripts, and profile.common is currently also run in the userShell, immediately before the shell specific scripts.
> That is in a bid to support simple activation scripts that syntactically evaluate in both bash and zsh to not need to be defined twice -- I have stated my concerns with this in https://github.com/flox/flox/pull/1149#discussion_r1528435769.

Thanks for highlighting - I've updated that issue thread with [my thoughts](https://github.com/flox/flox/pull/1149#discussion_r1530036389), and I'll clarify this point in the proposal above.  👍 

> Proposal 3.3.iii 🙃 asks to simply invoke exec ... although our current understanding is that CMD'ed commands run in virtually the same environment an interactive shell would provide.
> hence we currently do something like exec userShell -c "source profile_scripts; subshell on-activate-hook; exec cmd args

Yes, and that will not work because the userShell requires access to variables exported to the environment during the [userShell-agnostic] activation hook. There are many examples of where this is useful, but this is probably the most familiar one:
```
eval $(ssh-agent -s)
```

I don't mean to suggest that this is the best way to invoke `ssh-agent` specifically, but it is an example that will be familiar to most and I've seen this method used for _plenty_ of other "daemonizing" applications, any of which our users might want to invoke as part of an environment activation. For example, in the span of 30 seconds I found [this article](https://markgituma.medium.com/kubernetes-local-to-production-with-django-3-postgres-with-migrations-on-minikube-31f2baa8926e) which advises a user to invoke `eval $(minikube docker-env)` as part of setting up a developer environment, which is pretty much the exact use case I'm referring to.

**Comment by ghudgins** [comment]
@stahnma to follow-up on where this is in plan

**Comment by mkenigs** [comment]
> > I'm probably just not seeing the inter-dependencies you're seeing?
> 
> For example, I only came up with proposal 1 and 2 when I found it difficult to write up a description of proposal 3. Proposals 1 and 2 are really trivial in that regard, just clarifying terminology more than anything else.
> 
> Proposal 4 is admittedly a different thing, but one that came up in the context of the sorts of complexities that can arise in stage 3 so I thought it provided good context to help affirm the correctness of the 3-stage design. It's also less of a priority to be fair; if/when we get to the point of finishing 1-3 and want to copy 4 to a new issue and close this out then I'd be OK with that. 👍

I think there's enough complexity here that it would be easier to discuss if it was split up into individual user facing bugs. E.g.:
-  direnv + bash + fish doesn't work with a specific reproducer
- `flox activate -- cmd` has surprising behavior in scenario _fill_in_the_blank_

That will allow us to discuss the ultimate outcome rather than getting confused on which issue we're talking about or distracted by implementation details.

**Comment by limeytexan** [comment]
Definitely agree with the desire to enumerate all the problems, but I do not agree with the idea of attacking those problems independently. That's why I logged this issue to find a _holistic_ solution to address all of the problems, and splitting those focuses would be the opposite of that.

**Comment by limeytexan** [comment]
On Sunday morning I sat down to extract a summary of the problems in the issues mentioned above and found myself poring over the code to validate the logic of the refactor. In the process I started tinkering just to confirm the activation flow and was pleasantly surprised at how straightforward it was to make rapid progress. By noon the refactor was mostly complete, and after a few more hours tests were passing. All in all the work boiled down to migrating a small amount of functionality from rust to a new top-level `activate` script rendered to the environment itself, and then updating the unit tests to reflect the updated activation flow. I'll be logging a PR momentarily, and I'm expecting that it will close or pave the way to close out or readdress multiple issues which include the following:

* #1019
    * this was previously closed out with the introduction of the manifest `[profile]` section, and this refactor uses this same UX but turns around the implementation so that the `[hook]` section is invoked just once as the environment is first activated, and then the `[profile]` section is sourced with each new subshell thereafter
    * crucially, this provides the ability for actions in the `[hook]` section to set environment variables inherited/used by the `[profile]` section
* #637
    * the refactor provides the opportunity to reliably run starship in each new subshell without having to put it in the user's dotfile
* #688
    * having clearly delineated actions to be performed only once as the environment is first activated, the refactor paves the way for adding logic to monitor for and tear down processes associated with that top-level "activation PID"
* #1186
    * I see Zach has a fix in progress in 1277 but I also happened to trip across and fix this issue as well
* #1174 refers to https://discourse.flox.dev/t/setopt-command-not-found/837
    * this was closed 2 weeks back but cannot happen now as the refactor more clearly delineates the different scripts to be used for different shells
* #1173
    * this work opened the door for a followup PR to add fish and tcsh support which I'll also be logging later today
* #1252
    * caused by a collision between the `SET_PROMPT_SH` variable being shared by zsh and bash, fixed in the above fish/tcsh PR
* #1185
    * I didn't include the fix for this in the refactor but will log a PR for it shortly
    * this is also potentially related to Yannik's work in 1144
* #1135
    * done with the refactor
* #1130
    * it doesn't address this directly, but the refactor adds a new envrc file saved to the environment which can be used to address this
* #817
    * this was closed out in favor of 1130 but not sure was completely addressed (may also be related to 1144 and 1185)
